Source/WTF/wtf/cocoa/TypeCastsCocoa.h

```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
 #define checked_objc_cast checked_objc_castARC
 #endif
 
-template<typename T> inline T* checked_objc_cast(id object)
+template <typename ExpectedType>
+struct ObjCTypeCastTraits {
+public:
+    static bool isType(id object) { return [object isKindOfClass:[ExpectedType class]]; }
+
+    template <typename ArgType>
+    static bool isType(const ArgType *object) { return [object isKindOfClass:[ExpectedType class]]; }
+};
+
+#define SPECIALIZE_OBJC_TYPE_TRAITS(ClassName, ClassExpresssion) \
+namespace WTF { \
+template <> \
+class ObjCTypeCastTraits<ClassName> { \
+public: \
+    static bool isType(id object) { return [object isKindOfClass:(ClassExpresssion)]; } \
+    static bool isType(const NSObject *object) { return [object isKindOfClass:(ClassExpresssion)]; } \
+}; \
+}
+
+template <typename ExpectedType, typename ArgType>
+inline bool is_objc(ArgType * source)
+{
+    return source && ObjCTypeCastTraits<ExpectedType>::isType(source);
+}
+
+template<typename T> inline T *checked_objc_cast(id object)
+{
+    if (!object)
+        return nullptr;
+
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
+
+    return reinterpret_cast<T*>(object);
+}
+
+template<typename T, typename U> inline T *checked_objc_cast(U *object)
 {
     if (!object)
         return nullptr;
 
-    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION([object isKindOfClass:[T class]]);
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
 
     return reinterpret_cast<T*>(object);
 }
 
 // Use dynamic_objc_cast<> instead of checked_objc_cast<> when actively checking NS types,
-// similar to dynamic_cast<> in C++. Be sure to include a nil check.
+// similar to dynamic_cast<> in C++.
```

Comment:
I wish we could match the naming of downcast / dynamicDowncast more closely. Maybe:
is_objc => isObjC
checked_objc_cast => downcastObjC
dynamic_objc_cast => dynamicDowncastObjC

or if we like underscores:
is_objc
checked_objc_cast => downcast_objc
dynamic_objc_cast => dynamic_downcast_objc

That way we have a consistent pattern for describing whether the cast is crashes or returns null.

Reply:
Hmmmm... the goal is still that we should only use these for downcast, right? I guess the challenge is how to static_assert that?

Maybe we can static_assert? This seems to work:

```
@interface Base : NSObject {
}

@end

@interface Derived : Base {
}

@end

void f() {
    static_assert(!std::is_base_of_v<Derived, Base>);
    static_assert(std::is_base_of_v<Base, Derived>);
}

```

Reply:
oh, neat. Didn't realize `is_base_of_v` worked with Obj-C classes.

Reply:
Hm... that doesn't quite work because we have an actual "side-way" cast.

```
/Volumes/Data/safari-2/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/cocoa/TypeCastsCocoa.h:124:19: error: static assertion failed due to requirement 'std::is_base_of_v<id<WKObject>, _WKResourceLoadStatisticsFirstParty>'
  124 |     static_assert(std::is_base_of_v<U, T>);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:27:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdPartyInternal.h:26:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/APIResourceLoadStatisticsThirdParty.h:28:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/API/APIObject.h:35:
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:63:12: note: in instantiation of function template specialization 'WTF::checked_objc_cast<_WKResourceLoadStatisticsFirstParty, id<WKObject>>' requested here
   63 |     return checked_objc_cast<DestinationClass>(source);
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:68:12: note: in instantiation of function template specialization 'WebKit::checkedObjCCast<_WKResourceLoadStatisticsFirstParty, id<WKObject>>' requested here
   68 |     return checkedObjCCast<typename WrapperTraits<ObjectClass>::WrapperClass>(object.wrapper());
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:88:12: note: in instantiation of function template specialization 'WebKit::wrapper<API::ResourceLoadStatisticsFirstParty>' requested here
   88 |     return wrapper(object.get());
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:51:16: note: in instantiation of function template specialization 'WebKit::wrapper<API::ResourceLoadStatisticsFirstParty>' requested here
   51 |         return wrapper(API::ResourceLoadStatisticsFirstParty::create(data));
      |                ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:50:12: note: in instantiation of function template specialization 'WTF::createNSArray<const WTF::Vector<WebKit::ITPThirdPartyDataForSpecificFirstParty> &, (lambda at /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:50:60)>' requested here
   50 |     return createNSArray(_thirdParty->underFirstParties(), [] (auto& data) {
      |            ^
1 error generated.
```